### PR TITLE
Improve error handling in ShellTask

### PR DESCRIPTION
**Pull request by pradithya**
# TL;DR
Improve the error returned by `ShellTask`. As of now, `ShellTask` error message is not helpful since it doesn't capture the `stderr` from the subprocess. Additionally, the error thrown by `ShellTask` is currently classified as system error instead of user error.

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 
## Complete description

- Raise `FlyteRecoverableError` from `ShellTask` so that it's classified as user's error and the retry policy specified in the task is respected. 
- Improve the error message captured by `ShellTask`. As of now the error message generated by `ShellTask` in flyteconsole looks as follows. Notice that it doesn't contain any details of the failed command. 

```
[3/3] currentAttempt done. Last Error: SYSTEM::Traceback (most recent call last):

      File "/usr/local/lib/python3.8/site-packages/flytekit/exceptions/scopes.py", line 165, in system_entry_point
        return wrapped(*args, **kwargs)
      File "/usr/local/lib/python3.8/site-packages/flytekit/core/base_task.py", line 527, in dispatch_execute
        raise e
      File "/usr/local/lib/python3.8/site-packages/flytekit/core/base_task.py", line 524, in dispatch_execute
        native_outputs = self.execute(**native_inputs)
      File "/usr/local/lib/python3.8/site-packages/flytekit/extras/tasks/shell.py", line 220, in execute
        subprocess.check_call(gen_script, shell=True)
      File "/usr/local/lib/python3.8/subprocess.py", line 364, in check_call
        raise CalledProcessError(retcode, cmd)

Message:

    Command '
        export RUN_DATE=2023-07-04 &&
        make run
    ' returned non-zero exit status 2.

SYSTEM ERROR! Contact platform administrators.
```

With this change, users would be able to see `stdout` and `stderr` of the subprocess as part of the error message. 

## Tracking Issue
[https://github.com/flyteorg/flyte/issues/3559](https://github.com/flyteorg/flyte/issues/3559)

## Follow-up issue
N.A.

### CODE REVIEW

File: flytekit/extras/tasks/shell.py [hunk 1]
```diff
@@ -213,21 +238,23 @@ def execute(self, **kwargs) -> typing.Any:
...
+            error = (
+                f"Failed to Execute Script, return-code {returncode} \n"
+                f"Current directory contents: .\n-{fstr}\n"
+                f"StdOut: {stdout}\n"
+                f"StdErr: {stderr}\n"
             )
-            raise
+            logger.error(error)
+            # raise FlyteRecoverableException so that it's classified as user error and will be retried
+            raise FlyteRecoverableException(error)
```
Comment (fg91) [inline flytekit/extras/tasks/shell.py:257]: I think not every user error should be a recoverable one 🤔 

Take a look at [how `PythonFunctionTask` separates user vs system error scopes](https://github.com/flyteorg/flytekit/blob/982ffe6097dd6700d084ff4da276fe4315927e77/flytekit/core/python_function_task.py#L164):

```python
return exception_scopes.user_entry_point(self._task_function)(**kwargs)
```

Comment (fg91) [reply flytekit/extras/tasks/shell.py:257]: Right, of course, makes sense :+1:
This change is not really backwards compatible though, I wonder whether this is a dealbreaker 🤔

Comment (pradithya) [reply flytekit/extras/tasks/shell.py:257]: Yes, there is undoubtedly a different behavior change when a shell task throws an error.

Before: Error in shell task is considered system error thus retried N number of times, where N is [3 by default](https://github.com/flyteorg/flytepropeller/blob/6a14e7fbffe89786fb1d8cde22715f93c2f3aff5/pkg/controller/config/config.go#L97C1-L98C1). 

After: Error in shell task is considered user recoverable error. By default, it's not retried and will only be retried if users specify retry in the task's metadata. 

 However, the shell task didn't respect the retry strategy set by users, and I think it can be considered as bug.

Comment (pradithya) [reply flytekit/extras/tasks/shell.py:257]: @wild-endeavor your opinion will be very helpful here.

Comment (pingsutw) [reply flytekit/extras/tasks/shell.py:257]: +1, I think shell task should throw user recoverable error

File: flytekit/extras/tasks/shell.py [hunk 2]
```diff
@@ -98,6 +99,30 @@ def interpolate(
...
+    to the current process stdout so that the subprocess execution will not appear unresponsive
+
+    :param script: script to be executed
+    :type script: str
+    :return: tuple containing the process returncode, stdout, and stderr
+    :rtype: typing.Tuple[int, str, str]
+    """
+    process = subprocess.Popen(script, stdout=subprocess.PIPE, stderr=subprocess.PIPE, bufsize=0, shell=True, text=True)
+
+    # print stdout so that long-running subprocess will not appear unresponsive
```
Comment (pingsutw) [inline flytekit/extras/tasks/shell.py:116]: qq: Currently, the pod's log didn't show any stdout and stderr?

**Review by pingsutw — APPROVED** [review]

File: flytekit/extras/tasks/shell.py (hunk 2 above)
Comment (pradithya) [reply flytekit/extras/tasks/shell.py:116]: Currently it does. However this PR capture the stdout thus need to emit it manually.
